router_test.go

```diff
@@ -2695,6 +2698,87 @@ func TestRouterHandleMethodOptions(t *testing.T) {
 	}
 }
 
+func TestRouter_Routes(t *testing.T) {
+	type rr struct {
+		method string
+		path   string
+		name   string
+	}
+	var testCases = []struct {
+		name        string
+		givenRoutes []rr
+		expect      []rr
+	}{
+		{
+			name: "ok, multiple",
+			givenRoutes: []rr{
+				{method: http.MethodGet, path: "/static", name: "/static"},
+				{method: http.MethodGet, path: "/static/*", name: "/static/*"},
+			},
+			expect: []rr{
+				{method: http.MethodGet, path: "/static", name: "/static"},
+				{method: http.MethodGet, path: "/static/*", name: "/static/*"},
+			},
+		},
+		{
+			name:        "ok, no routes",
+			givenRoutes: []rr{},
+			expect:      []rr{},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			dummyHandler := func(Context) error { return nil }
+
+			e := New()
+			route := e.router
+
+			for _, tmp := range tc.givenRoutes {
+				route.add(tmp.method, tmp.path, tmp.name, dummyHandler)
+			}
+
+			// Add does not add route. because of backwards compatibility we can not change this method signature
+			route.Add("LOCK", "/users", handlerFunc)
```

Comment:
Could you detail why `Add()` does not add a route?
If it is just a test for backward compatibility it should not be in the testcase loop here I guess.

Reply:
because `Add` signature is `func (r *Router) Add(method, path string, h HandlerFunc) {`
and `add` signature is `func (r *Router) add(method, path, name string, h HandlerFunc) *Route {` and I needed way to return routes from Router.

`add` was introduced so we could create `Route` object inside router. As router should be the creator of routes not Echo object. Previously it was done inside `echo.add` method and cause problems when multiple `host`s are in play
https://github.com/labstack/echo/blob/8d4ac4c907696c1aecb237d088f9b648c00f274b/echo.go#L530-L545

and this `Add` call in this test is to document that behavior.

Reply:
So to summarize: This test ensures the routes added via Add() do not show up in non-global routers.

Maybe you can improve the comment. I had to lookup the code 
Suggestion (please verify if it makes sense):

```diff
-                       // Add does not add route. because of backwards compatibility we can not change this method signature
+                       // Verify backwards compatibility for documented behavior for router.Add() 
+                       // Routes added via Add() shall not show up for dedicated routers (e.g. used for host based routes)
 			route.Add("LOCK", "/users", handlerFunc)
```